File: skills/kb-lint/scripts/conformance.py
#!/usr/bin/env python3
"""Deterministic OKF v0.1 conformance check for a knowledge bundle (§9).

Usage:  python3 conformance.py <bundle-dir> [--json]

Exit code is non-zero if any ERROR is present. Broken links and soft-guidance
issues are reported as WARN and never fail (SPEC §5.3 / §9 — consumers MUST
tolerate them). This checks structure only; drift (contradictions, stale
claims, orphans, coverage gaps) is the fuzzy, agent-driven half of kb-lint.
"""
import os, re, sys, json, posixpath

RESERVED = {"index.md", "log.md"}
FM_RE = re.compile(r"^---\n(.*?)\n---\n?", re.S)
HEADING_RE = re.compile(r"^#{1,6}\s+(.+?)\s*$", re.M)
LINK_RE = re.compile(r"\]\(([^)#\s]+\.md)(#[^)]*)?\)")
LOG_DATE_RE = re.compile(r"^##\s+(.+?)\s*$", re.M)
ISO_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def frontmatter(text):
    m = FM_RE.match(text)
    return m.group(1) if m else None


def check(bundle):
    errors, warns = [], []
    md = []
    for dp, _, fs in os.walk(bundle):
        for f in fs:
            if f.endswith(".md"):
                md.append(os.path.relpath(os.path.join(dp, f), bundle))

    for rel in sorted(md):
        text = open(os.path.join(bundle, rel), encoding="utf-8").read()
        base = os.path.basename(rel)
        fm = frontmatter(text)

        if base in RESERVED:
            # Reserved files carry no frontmatter, except the ROOT index.md may
            # declare okf_version (SPEC §6/§11).
            if fm is not None:
                is_root_index = (rel == "index.md")
                if not (is_root_index and "okf_version" in fm):
                    errors.append(f"{rel}: reserved file must not carry frontmatter")
            if base == "log.md":
                for m in LOG_DATE_RE.finditer(text):
                    if not ISO_DATE_RE.match(m.group(1)):
                        warns.append(f"{rel}: log date heading not ISO 8601: '{m.group(1)}'")
            continue

        # Concept document: rules 1 & 2.
        if fm is None:
            errors.append(f"{rel}: concept has no parseable frontmatter")
            continue
        tm = re.search(r"^type:\s*(.+?)\s*$", fm, re.M)
        if not tm or not tm.group(1).strip():
            errors.append(f"{rel}: missing or empty required 'type'")

    # Broken relative links → WARN only (never a conformance failure).
    for rel in md:
        srcdir = posixpath.dirname(rel)
        for m in LINK_RE.finditer(open(os.path.join(bundle, rel), encoding="utf-8").read()):
            tgt = m.group(1)
            if "://" in tgt:
                continue
            resolved = posixpath.normpath(posixpath.join(srcdir, tgt)) if not tgt.startswith("/") \
                else tgt.lstrip("/")
            if not os.path.exists(os.path.join(bundle, resolved)):
                warns.append(f"{rel}: broken link -> {tgt}")

    return {"bundle": bundle, "concepts": len([f for f in md if os.path.basename(f) not in RESERVED]),
            "files": len(md), "errors": errors, "warnings": warns}


def main(argv):
    args = [a for a in argv[1:] if not a.startswith("--")]
    as_json = "--json" in argv
    bundle = args[0] if args else "."
    if not os.path.isdir(bundle):
        print(f"not a directory: {bundle}", file=sys.stderr)
        return 2
    r = check(bundle)
    if as_json:
        print(json.dumps(r, indent=2))
    else:
        print(f"{r['bundle']}: {r['files']} files, {r['concepts']} concepts")
        for e in r["errors"]:
            print(f"  ERROR  {e}")
        for w in r["warnings"]:
            print(f"  warn   {w}")
        verdict = "CONFORMANT" if not r["errors"] else "NON-CONFORMANT"
        print(f"  => {verdict} ({len(r['errors'])} errors, {len(r['warnings'])} warnings)")
    return 1 if r["errors"] else 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: skills/kb-lint/scripts/conformance.mjs
#!/usr/bin/env node

// packages/kb-tools/src/conformance.ts
import { existsSync, readFileSync, statSync as statSync2 } from "node:fs";
import { dirname, join as join2 } from "node:path";

// packages/kb-tools/src/shared.ts
import { readdirSync, statSync } from "node:fs";
import { join, relative, sep } from "node:path";
var FM_RE = /^---\n([\s\S]*?)\n---\n?/;
var RESERVED = /* @__PURE__ */ new Set(["index.md", "log.md"]);
function collectMarkdown(bundle) {
  const out = [];
  const walk = (dir) => {
    let entries;
    try {
      entries = readdirSync(dir).sort();
    } catch {
      return;
    }
    for (const name of entries) {
      const full = join(dir, name);
      let st;
      try {
        st = statSync(full);
      } catch {
        continue;
      }
      if (st.isDirectory()) {
        walk(full);
      } else if (name.endsWith(".md")) {
        out.push(relative(bundle, full).split(sep).join("/"));
      }
    }
  };
  walk(bundle);
  return out;
}
function frontmatter(text) {
  const m = FM_RE.exec(text);
  return m ? m[1] : null;
}
var NON_ASCII = new RegExp("[" + String.fromCharCode(128) + "-" + String.fromCharCode(65535) + "]", "g");
function pythonJson(value) {
  return JSON.stringify(value, null, 2).replace(
    NON_ASCII,
    (c) => "\\u" + c.charCodeAt(0).toString(16).padStart(4, "0")
  );
}
function normalizePosix(p) {
  const isAbs = p.startsWith("/");
  const parts = p.split("/");
  const stack = [];
  for (const part of parts) {
    if (part === "" || part === ".") continue;
    if (part === "..") {
      if (stack.length && stack[stack.length - 1] !== "..") stack.pop();
      else if (!isAbs) stack.push("..");
    } else {
      stack.push(part);
    }
  }
  const joined = stack.join("/");
  if (isAbs) return "/" + joined;
  return joined === "" ? "." : joined;
}

// packages/kb-tools/src/conformance.ts
var HEADING_LOG_RE = /^##\s+(.+?)\s*$/gm;
var ISO_DATE_RE = /^\d{4}-\d{2}-\d{2}$/;
var LINK_RE = /\]\(([^)#\s]+\.md)(#[^)]*)?\)/g;
var TYPE_RE = /^type:\s*(.+?)\s*$/m;
function checkConformance(bundle) {
  const errors = [];
  const warnings = [];
  const md = collectMarkdown(bundle);
  const posixBasename = (rel) => rel.split("/").pop() ?? rel;
  for (const rel of [...md].sort()) {
    const text = readFileSync(join2(bundle, rel), "utf-8");
    const base = posixBasename(rel);
    const fm = frontmatter(text);
    if (RESERVED.has(base)) {
      if (fm !== null) {
        const isRootIndex = rel === "index.md";
        if (!(isRootIndex && fm.includes("okf_version"))) {
          errors.push(`${rel}: reserved file must not carry frontmatter`);
        }
      }
      if (base === "log.md") {
        for (const m of text.matchAll(HEADING_LOG_RE)) {
          if (!ISO_DATE_RE.test(m[1])) {
            warnings.push(`${rel}: log date heading not ISO 8601: '${m[1]}'`);
          }
        }
      }
      continue;
    }
    if (fm === null) {
      errors.push(`${rel}: concept has no parseable frontmatter`);
      continue;
    }
    const tm = TYPE_RE.exec(fm);
    if (!tm || !tm[1].trim()) {
      errors.push(`${rel}: missing or empty required 'type'`);
    }
  }
  for (const rel of md) {
    const srcdir = dirname(rel) === "." ? "" : dirname(rel);
    const text = readFileSync(join2(bundle, rel), "utf-8");
    for (const m of text.matchAll(LINK_RE)) {
      const tgt = m[1];
      if (tgt.includes("://")) continue;
      const resolved = tgt.startsWith("/") ? tgt.replace(/^\/+/, "") : normalizePosix(srcdir ? `${srcdir}/${tgt}` : tgt);
      if (!existsSync(join2(bundle, resolved))) {
        warnings.push(`${rel}: broken link -> ${tgt}`);
      }
    }
  }
  return {
    bundle,
    concepts: md.filter((f) => !RESERVED.has(posixBasename(f))).length,
    files: md.length,
    errors,
    warnings
  };
}
function formatReport(r) {
  const lines = [`${r.bundle}: ${r.files} files, ${r.concepts} concepts`];
  for (const e of r.errors) lines.push(`  ERROR  ${e}`);
  for (const w of r.warnings) lines.push(`  warn   ${w}`);
  const verdict = r.errors.length === 0 ? "CONFORMANT" : "NON-CONFORMANT";
  lines.push(`  => ${verdict} (${r.errors.length} errors, ${r.warnings.length} warnings)`);
  return lines.join("\n");
}
function runCli(argv) {
  const args = argv.filter((a) => !a.startsWith("--"));
  const asJson = argv.includes("--json");
  const bundle = args[0] ?? ".";
  let isDir = false;
  try {
    isDir = statSync2(bundle).isDirectory();
  } catch {
    isDir = false;
  }
  if (!isDir) {
    process.stderr.write(`not a directory: ${bundle}
`);
    return 2;
  }
  const r = checkConformance(bundle);
  if (asJson) {
    process.stdout.write(pythonJson(r) + "\n");
  } else {
    process.stdout.write(formatReport(r) + "\n");
  }
  return r.errors.length ? 1 : 0;
}

// packages/kb-tools/src/cli/conformance-cli.ts
process.exit(runCli(process.argv.slice(2)));
